spectree/plugins/falcon_plugin.py

```diff
@@ -32,20 +35,33 @@
 
 
 class StreamWrapper:
-    def __init__(self, stream):
+    def __init__(self, stream: BufferedReader):
         self._buf = CachedFile()
-        self._buf.write(stream)
+        stream.pipe(self._buf)
         self._buf.seek(0)
 
     def read(self, size: Optional[int] = -1, /) -> bytes:
-        return self._buf.read(size)
+        """read bytes from the stream, size -1 or None means max bytes"""
+        return self._buf.read(size if size is not None else -1)
 
     def exhaust(self) -> None:
         self._buf.seek(0)
         self._buf.truncate(0)
 
 
 class AsyncStreamWrapper(StreamWrapper):
+    def __init__(self):
+        self._buf = CachedFile()
+
+    @classmethod
+    async def from_stream(cls, stream: ASGIBufferedReader):
+        obj = cls()
+        loop = asyncio.get_running_loop()
+        async for chunk in stream:
+            await loop.run_in_executor(None, obj._buf.write, chunk)
+        obj._buf.seek(0)
```

Comment:
Probably more a theoretical edge case, but seeking should also be run in an executor.

It can block, especially on network file systems, or essentially any kind of storage if the file is large enough so that the beginning of it is not cached anymore, although I guess practically on an SSD the corresponding syscall will be blazing fast.

At least `aiofiles` seems to delegate this to an executor as well: https://github.com/Tinche/aiofiles/blob/main/src/aiofiles/threadpool/binary.py#L42.